Updating credit checkout page design

tasawernawaz opened the pull request:
Updating styling and structure of credit checkout page according to new design.

Please review this. @awais786 @ahsan-ul-haq  @aamir-khan @zubair-arbi 

Screenshots are attached for desktop, tablet and mobile respectively.

<img width="1069" alt="desktop" src="https://cloud.githubusercontent.com/assets/4245618/8931752/f62015fc-3551-11e5-91b7-f010f316df4d.png">
<img width="597" alt="tablet" src="https://cloud.githubusercontent.com/assets/4245618/8931754/f64aca0e-3551-11e5-9ce0-c7f17fb04b6b.png">
<img width="384" alt="mobile" src="https://cloud.githubusercontent.com/assets/4245618/8931753/f6269756-3551-11e5-8b66-ccfd89b8857f.png">

zubair-arbi commented on ecommerce/static/sass/views/_credit.scss:
```diff
@@ -56,18 +48,21 @@
 
     .payment-button {
-      width: 100%;
+      background-color: #0b6b0d;
```
Please use default colors of the theme from variables.

zubair-arbi commented on ecommerce/static/sass/views/_credit.scss:
```diff
@@ -77,13 +72,34 @@
+    .row {
+      .col-sm-4 {
+        .advantages {
+          border-top: 15px solid #f5f5f5;
```
Please use default colors of the theme from variables.

zubair-arbi commented on ecommerce/static/sass/views/_credit.scss:
```diff
@@ -77,13 +72,34 @@
+     width: 100%;
+     margin-bottom: 2px;
+  }
+}
```
New line at the end of file. There is setting in pycharm which automatically adds new line at the end of file.

zubair-arbi commented on ecommerce/templates/edx/credit/checkout.html:
```diff
@@ -5,21 +5,74 @@
-                        You are purchasing 1 credit hour for:
-                    {% plural %}
-                        You are purchasing {{ hours }} credit hours for:
+    <div class="container credit-checkout" xmlns="http://www.w3.org/1999/html">
```
Don't need to add `xmlns="http://www.w3.org/1999/html"`

zubair-arbi commented on ecommerce/templates/edx/credit/checkout.html:
```diff
@@ -5,21 +5,74 @@
+            <div class="col-sm-11">
+                <h3 class="title">
+                    {% blocktrans with provider='<span class="provider-name"></span>' count hours=credit_seats.0.attr.credit_hours %}
+                        Purchase 1 credit from {{ provider }} for:
```
Either remove hard-coded credit hours used in `Purchase 1 credit from`.

tasawernawaz replied:
In tag "blocktrans" here, this will only render when there is only one credit.

zubair-arbi commented on ecommerce/templates/edx/credit/checkout.html:
```diff
@@ -43,49 +98,14 @@ <h4 class="header-title title">
+        </div>
+        <div class="row center-block">
+            {% blocktrans with link='<a href="">our FAQs to view common quistions about our certificates</a>' %}
+                Please read {{ link }}
```
nitpick: Add dot at the sentence end.

zubair-arbi commented:
There is one test failing in `credit/tests/test_views.py` Please fix it. Also it would be nice If you add screen shots of different views (tablet, mobile etc)

zubair-arbi commented:
Great work :+1:

tasawernawaz commented:
@wedaly @AlasdairSwan Please review this PR, I have attached screenshots for desktop, tablet and mobile devices.

clintonb commented:
You didn't write it; but, since you're touching this area of the code, please remove `#provider-details-tpl` and use require JS to load templates. See https://github.com/edx/ecommerce/blob/master/ecommerce/static/js/views/course_detail_view.js for an example of doing this with the `text!` directive.

AlasdairSwan commented on ecommerce/static/sass/views/_credit.scss:
```diff
@@ -77,13 +73,34 @@
-    .col-xs-6 {
-      width: initial;
+/* Small devices (tablets) */
+@media (max-width: 992px) {
```
Can you please update this file to use the mobile first approach utilizing min-width (here is a good overview http://www.zell-weekeat.com/how-to-write-mobile-first-css/ )
Can you please set the widths as variables in the variables partial for re-usability https://github.com/edx/ecommerce/blob/master/ecommerce/static/sass/utilities/_variables.scss
768px is a pretty common breakpoint size but I'm curious where you got 992px from? Is it a Bootstrap breakpoint or is that when the layout starts to break?

tasawernawaz replied:
@AlasdairSwan I will change it from max to min (mobile first approach). Here is a link link for bootstrap media quarries. http://getbootstrap.com/css/#grid-media-queries

clintonb replied:
There are SASS variables for these widths. Please use them instead of hardcoding these values. https://github.com/twbs/bootstrap-sass/blob/master/assets/stylesheets/bootstrap/_variables.scss#L286

AlasdairSwan commented on ecommerce/static/js/views/provider_view.js:
```diff
@@ -22,6 +22,7 @@ define([
                     success: function(response) {
                         if (response.length) {
                             var html = self.template(response[0]); //currently we are assuming only one provider
+                            $('.title').find('.provider-name').text(response[0]['display_name']);
```
If the title element not in this template? I am concerned with the loose scoping of this, .title seems to be not a terribly specific class. It's also not in keeping with the Backbone MV* approach if the title element is in another view in that it is binding this view to another. It would be better to have the other view listen for changes to the model and update based on that.

clintonb commented:
@AlasdairSwan and I ran into issues while trying to test this locally. It turns out they were due to the fact that we had not configured credit eligibility in LMS. It would be nice if this were mentioned in the error message (e.g. "You are not eligible...") rather than the generic message "Something went wrong".

When Support inevitable gets a call about an error on this page, a more specific error message will help us quickly debug the problem.

clintonb commented:
Also, while testing this locally I was able to view the page despite not having eligibility data in LMS. This is worth investigating to ensure it doesn't happen in production.

tasawernawaz commented:
I have updated branch for addressing comments and used backbone collections and update views.

Please review this @zubair-arbi @awais786  @ahsan-ul-haq @aamir-khan

tasawernawaz commented:
@clintonb There are tow API calls. One for provider detail and 2nd is for user eligibility, Can you please suggest error messages for both in can of failures ?

tasawernawaz commented:
@AlasdairSwan @wedaly Please review this.

AlasdairSwan commented on ecommerce/static/js/collections/provider-collection.js:
```diff
@@ -0,0 +1,14 @@
+
+    var providerModel = Backbone.Model.extend({
+
+    });
```
Please move the model to its own file that you can pass in via the define statement.

AlasdairSwan commented on ecommerce/static/js/collections/user-eligibility-collection.js:
```diff
@@ -0,0 +1,15 @@
+
+    var eligibilityModel = Backbone.Model.extend({
+
+    });
```
Please move the model to its own file that you can pass in via the define statement.

AlasdairSwan commented on ecommerce/static/js/views/provider_view.js:
```diff
@@ -1,68 +1,50 @@
-                            self.toggleProviderContent(true);
-                        } else {
+
+                this.collection.url += "?provider_id=" + this.$el[0].dataset.providersIds;
```
It would be cleaner if the collection had a setUrl function that you pass the provider id(s) to.

AlasdairSwan commented on ecommerce/static/js/views/provider_view.js:
```diff
@@ -1,68 +1,50 @@
-                    }
-                })
+            renderProviderDetail: function (collecttion) {
+                var providerData = collecttion.toJSON();
```
Spelling error: it should be 'collection'.

AlasdairSwan commented on ecommerce/static/js/views/provider_view.js:
```diff
@@ -1,68 +1,50 @@
+                if (providerData.length) {
+                    // Currently we are assuming that we are having only one provider
+                    $('.title').find('.provider-name').text(providerData[0]['display_name']);
+                    var template = _.template(providerTemplate);
```
All variables should be set at the top of the function (because hoisting).

AlasdairSwan commented on ecommerce/static/js/views/user_eligibility_view.js:
```diff
@@ -0,0 +1,54 @@
+        });
+    });
+
+
```
Inconsistent quotes (sometimes " and sometimes '). Please stick to single quotes throughout.

AlasdairSwan commented on ecommerce/static/js/views/user_eligibility_view.js:
```diff
@@ -0,0 +1,54 @@
+                var eligibilityData = collection.toJSON();
+                if (eligibilityData.length) {
+                    var deadline = new Date(eligibilityData[0]["deadline"]),
+                        formattedDate = deadline.toLocaleString('en-us', {month: "long"}) + " " + deadline.getDay() + "," + deadline.getFullYear();
```
All variables should be set at the top of the function (because hoisting - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/var#var_hoisting).

tasawernawaz commented:
@AlasdairSwan I have updated as per your suggestions, kindly review it.

AlasdairSwan commented on ecommerce/static/js/views/user_eligibility_view.js:
```diff
@@ -0,0 +1,56 @@
+                $('.error-message').toggleClass('hide', isEnabled);
+            }
+        });
+    });
```
The ); should be on a new line i.e.
```
    }
);
```

AlasdairSwan commented on ecommerce/static/js/views/provider_view.js:
```diff
@@ -1,68 +1,51 @@
             }
         });
-});
+    });
```
The ); should be on a new line i.e.
```
    }
);
```

AlasdairSwan commented on ecommerce/static/js/models/user-eligibility-model.js:
```diff
@@ -0,0 +1,11 @@
+    });
+
+});
+
```
Please update spacing to be consistent with other Backbone file (e.g. ecommerce/static/js/views/user_eligibility_view.js)

AlasdairSwan commented on ecommerce/static/js/models/provider-model.js:
```diff
@@ -0,0 +1,11 @@
+    });
+
+});
+
```
Please update spacing to be consistent with other Backbone file (e.g. ecommerce/static/js/views/user_eligibility_view.js)

AlasdairSwan commented on ecommerce/static/js/collections/user-eligibility-collection.js:
```diff
@@ -0,0 +1,16 @@
+        }
+    });
+});
+
```
Please update spacing to be consistent with other Backbone file (e.g. ecommerce/static/js/views/user_eligibility_view.js)

AlasdairSwan commented on ecommerce/static/js/collections/provider-collection.js:
```diff
@@ -0,0 +1,14 @@
+                this.url += '?provider_ids=' + providerIds;
+        }
+    });
+});
```
Please update spacing to be consistent with other Backbone file (e.g. ecommerce/static/js/views/user_eligibility_view.js)

AlasdairSwan commented on ecommerce/static/js/views/provider_view.js:
```diff
@@ -1,68 +1,51 @@
-                })
+            renderProviderDetail: function (collection) {
+                var providerData = collection.toJSON(),
+                    template =  _.template(providerTemplate);
```
Although template should be declared here it should not be set unless there is a providerData.length. (see the deadline var in renderEligibilityDate() in ecommerce/static/js/views/user_eligibility_view.js

AlasdairSwan commented:
Once my layout comments have been addressed and the conflicts have been resolved it is good to go :+1:

srpearce commented:
@tasawernawaz Sorry, I missed your comment yesterday. Can you tell me what the error messages currently say and give me a little context - why would the errors occur, and would we want to give users specific information? IOW, can we just say something generic like "An error has occurred. Wait a few minutes and try again."?

tasawernawaz commented:
@srpearce There are two different APIs on credit checkout page. 1st one is responsible for getting providers detail and 2nd one is for checking eligibility and deadline date against user. 
Currently I am showing generic message "Something went wrong! Please try again later." But in future I think we can show more specific error messages on failure or no data found (e.g user is not eligible or there is no provider to show).
